### ISSUE TITLE
Loss curve differs when using cache dataset 

### ISSUE THREAD

**Issue by JiaxinZhuang** [Project-MONAI/MONAI#5331]
**Describe the bug**
Hi, when I use the cache dataset in MONAI, it indeed improves the training speed. However, compared with the code without using cache dataset, the loss decreases slower. Based on my understanding of cache dataset's mechanism, this seems not reasonable. They should have a similar loss value on the same epoch. For the code without cache dataset, I run it twice time and they have very similar loss curves.

**To Reproduce**
The code below is used for generating dataset.
```

  train_transforms = Compose(
        [
            LoadImaged(keys=["image"]),
            AddChanneld(keys=["image"]),
            Orientationd(keys=["image"], axcodes="RAS"),
            ScaleIntensityRanged(
                keys=["image"], a_min=args.a_min, a_max=args.a_max, b_min=args.b_min, b_max=args.b_max, clip=True
            ),
            SpatialPadd(keys="image", spatial_size=[args.roi_x, args.roi_y, args.roi_z]),
            CropForegroundd(keys=["image"], source_key="image", k_divisible=[args.roi_x, args.roi_y, args.roi_z]),
            RandSpatialCropSamplesd(
                keys=["image"],
                roi_size=[args.roi_x, args.roi_y, args.roi_z],
                num_samples=args.sw_batch_size,
                random_center=True,
                random_size=False,
            ),
            ToTensord(keys=["image"]),
        ]
    )
  
    if args.cache_dataset:
        print("Using MONAI Cache Dataset")
        train_ds = CacheDataset(data=datalist, transform=train_transforms, cache_rate=1.0, num_workers=num_workers)
    else:
        print("Using generic dataset")
        train_ds = Dataset(data=datalist, transform=train_transforms)

    if args.distributed:
        train_sampler = DistributedSampler(dataset=train_ds, even_divisible=True, shuffle=True)
    else:
        train_sampler = None
    train_loader = DataLoader(
        train_ds, batch_size=args.batch_size, num_workers=num_workers, sampler=train_sampler, drop_last=True, pin_memory=True, persistent_workers=False
    )

```

**Screenshots**
<img width="1096" alt="image" src="https://user-images.githubusercontent.com/25994409/195787413-2998759b-327a-4c40-80eb-fb84409dfcb8.png">

**Environment**
Ubuntu
```
torch==1.8.1+cu111
torchvision==0.9.1+cu111
monai==1.0.0
```

**Comment by rijobro** [comment]
Do you use seeds? How do you know that you're not just visualising one bad run with `CacheDataset` vs one good run with `Dataset`?

If the problem isn't seeds, can you make a minimum working example using open-source data?
